crates/sui-bridge-indexer/src/eth_worker.rs

```diff
@@ -1,21 +1,162 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
+use crate::config::Config;
+use crate::latest_eth_syncer::LatestEthSyncer;
+use crate::metrics::BridgeIndexerMetrics;
+use crate::postgres_manager::get_latest_eth_token_transfer;
 use crate::postgres_manager::{write, PgPool};
 use crate::{BridgeDataSource, TokenTransfer, TokenTransferData, TokenTransferStatus};
+use anyhow::Result;
 use ethers::providers::Provider;
 use ethers::providers::{Http, Middleware};
 use ethers::types::Address as EthAddress;
+use mysten_metrics::spawn_logged_monitored_task;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::str::FromStr;
 use std::sync::Arc;
 use sui_bridge::abi::{EthBridgeEvent, EthSuiBridgeEvents};
 use sui_bridge::types::EthLog;
+use sui_bridge::{eth_client::EthClient, eth_syncer::EthSyncer};
+use tokio::task::JoinHandle;
 use tracing::info;
 use tracing::log::error;
 
-pub async fn process_eth_events(
-    mut eth_events_rx: mysten_metrics::metered_channel::Receiver<(EthAddress, u64, Vec<EthLog>)>,
+#[derive(Clone, Debug)]
+pub struct EthBridgeWorker {
+    provider: Arc<Provider<Http>>,
+    pg_pool: PgPool,
+    metrics: BridgeIndexerMetrics,
+    bridge_address: EthAddress,
+    config: Config,
+}
+
+impl EthBridgeWorker {
+    pub fn new(pg_pool: PgPool, metrics: BridgeIndexerMetrics, config: Config) -> Self {
```

Comment:
do we want to return Result<Self> here and let `main` panic?

Reply:
it means we do not want to panic inside the function, instead return an Error to the caller to handle.
The change does not achieve this purpose